### PR TITLE
Decompiler feature: remove redundant else scopes

### PR DESCRIPTION
**Pull request by intranautic**

### CODE REVIEW

File: angr/analyses/decompiler/optimization_passes/flip_boolean_cmp.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            if isinstance(node.false_node, SequenceNode):
+                for node in node.false_node.nodes:
+                    if not isinstance(node, Block):
+                        break
+                else:
+                    last_block = node
+            elif isinstance(node.false_node, Block):
+                last_block = node.false_node
+
+            if last_block and not [*self._graph.successors(last_block)]:
```
Comment (mahaloz) [inline angr/analyses/decompiler/optimization_passes/flip_boolean_cmp.py:39]: Don't do this unpacking. Explicitly convert generators to lists if you are using it like a list.

File: angr/analyses/decompiler/optimization_passes/flip_boolean_cmp.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from ..structuring.structurer_nodes import ConditionNode, SequenceNode
+from .optimization_pass import SequenceOptimizationPass, OptimizationPassStage
+
+
+class FlipBooleanCmp(SequenceOptimizationPass):
+    ARCHES = ["X86", "AMD64"]
+    PLATFORMS = ["linux", "windows", "cgc"]
+    STAGE = OptimizationPassStage.AFTER_STRUCTURING
+    NAME = "flip boolean comparison"
+    DESCRIPTION = "there is no salvation for the wicked, free will is an illusion"
```
Comment (mahaloz) [inline angr/analyses/decompiler/optimization_passes/flip_boolean_cmp.py:15]: Update this.

File: angr/analyses/decompiler/optimization_passes/__init__.py [hunk 3]
```diff
@@ -17,7 +17,7 @@
 from .register_save_area_simplifier import RegisterSaveAreaSimplifier
 from .ret_addr_save_simplifier import RetAddrSaveSimplifier
 from .x86_gcc_getpc_simplifier import X86GccGetPcSimplifier
-
+from .flip_boolean_cmp import FlipBooleanCmp
```
Comment (mahaloz) [inline angr/analyses/decompiler/optimization_passes/__init__.py:20]: Put a newline after this line

File: angr/analyses/decompiler/structured_codegen/c.py [hunk 4]
```diff
@@ -2767,9 +2780,27 @@ def _handle_Condition(self, condition_node: ConditionNode, **kwargs):
 
         else_node = self._handle(condition_node.false_node, is_expr=False) if condition_node.false_node else None
 
+        true_node = condition_node.true_node
+        do_simplify_else = False
+
+        if self.simplify_else_scope:
+            last_block = None
```
Comment (mahaloz) [inline angr/analyses/decompiler/structured_codegen/c.py:2787]: Make this entire block a boolean method in `c.py` (put it near the to). Maybe call it:
```python
def is_simple_return_node(node: Union[ailment.Block, SequenceNode]) -> bool:
    ...
```

**Review by mahaloz — CHANGES_REQUESTED** [review]

File: angr/analyses/decompiler/decompiler.py [hunk 5]
```diff
@@ -243,6 +243,7 @@ def progress_callback(p, **kwargs):
             self.func,
             seq_node,
             cfg=self._cfg,
+            clinic=clinic,
```
Comment (ltfish) [inline angr/analyses/decompiler/decompiler.py:246]: Let's pass in the output that Clinic generates, instead of passing in the entire Clinic instance.

**Comment by mahaloz** [comment]
@intranautic don't forget to do two more things:
1. Add Decompiler option so this can be disabled (and test it)
2. Add a testcase for `od.o`. This will require you to make a second PR in the binaries repo, which I can show you how to do.

**Review by mahaloz — APPROVED** [review]
Includes a new test case for this feature! Nice job!

File: angr/analyses/decompiler/optimization_passes/flip_boolean_cmp.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class FlipBooleanCmp(SequenceOptimizationPass):
+    """
+    In the scenario in which a false node has no apparent successors, flip the condition on that if-stmt.
+    This is only useful when StructuredCodeGenerator has simplify_else_scopes enabled, as this will allow the
+    flipped if-stmt to remove the redundant else.
+    """
+
+    ARCHES = ["X86", "AMD64"]
```
Comment (ltfish) [inline angr/analyses/decompiler/optimization_passes/flip_boolean_cmp.py:17]: Why are we limiting architectures to x86 and amd64? Also why are we limiting platforms? This seems to be a generic optimization for all architectures and platforms.

**Review by ltfish — CHANGES_REQUESTED** [review]
Please remove arch and platform limits or justify them.

**Review by ltfish — APPROVED** [review]

**Pull request merged**